File: src/net/merayen/elastic/ui/objects/contextmenu/Menu.java
package net.merayen.elastic.ui.objects.contextmenu;

import java.util.ArrayList;
import java.util.List;

import net.merayen.elastic.ui.UIObject;

class Menu extends UIObject {
	float radius = 150f;

	private float pointer_x;
	private float pointer_y;

	private int selected = -1;

	private List<ContextMenuItem> items = new ArrayList<>();
	private final int count;

	Menu(int count) {
		this.count = count;
	}

	@Override
	protected void onDraw() {
		int steps = count;

		draw.setColor(200, 200, 200);
		draw.setStroke(10);
		draw.oval(0, 0, radius * 2, radius * 2);

		float item_radius = radius / (3 * (steps / (float)8)) ;
		boolean marked = false;

		selected = -1;

		for(int i = 0; i < steps; i++) {
			boolean active = false;
			float x = radius - item_radius + (float)Math.sin(i / (float)steps * Math.PI * 2) * radius;
			float y = radius - item_radius + (float)Math.cos(i / (float)steps * Math.PI * 2) * radius;

			if(!marked && (Math.abs(pointer_x) > radius / 3 || Math.abs(pointer_y) > radius / 3)) {
				double pointer = (((Math.atan2(-pointer_x, -pointer_y) / Math.PI + 1) / 2) * steps + 0.5f) % steps;
				if((int)pointer == i) {
					marked = true;
					active = true;
				}
			}

			int menu_index = Math.floorMod(-i + count / 2, steps); // Makes items begin at 12 o'clock
			if(menu_index < items.size()) {
				ContextMenuItem cmi = items.get(menu_index);
				cmi.active = active;
				cmi.item_radius = item_radius;
				cmi.translation.x = x;
				cmi.translation.y = y;
				if(active)
					selected = menu_index;
			}
		}
	}

	void setPointer(float x, float y) {
		pointer_x = x;
		pointer_y = y;
	}

	void addMenuItem(ContextMenuItem item) {
		items.add(item);
		add(item);
	}

	ContextMenuItem getSelected() {
		if(selected > -1 && selected < items.size())
			return items.get(selected);

		return null;
	}
}

File: src/main/kotlin/net/merayen/elastic/ui/objects/contextmenu/Menu.kt
package net.merayen.elastic.ui.objects.contextmenu

import net.merayen.elastic.ui.MutableColor
import net.merayen.elastic.ui.Draw
import net.merayen.elastic.ui.UIObject
import java.util.*
import kotlin.math.min

internal class Menu(private val count: Int) : UIObject() {
    var radius = 150f
    var selectionRadius = 50f
    var backgroundColor = MutableColor(0.1f, 0.1f, 0.1f)

    private var currentRadius = 10f // For animation

    private var pointerX: Float = 0.toFloat()
    private var pointerY: Float = 0.toFloat()

    private var selected = -1

    private val items = ArrayList<ContextMenuItem>()

    private var lastDraw = System.currentTimeMillis()

    override fun onDraw(draw: Draw) {
        val steps = count

        currentRadius += (radius - currentRadius) / ((System.currentTimeMillis() - lastDraw) / 1000f) / 200f
        currentRadius = min(currentRadius, radius)
        lastDraw = System.currentTimeMillis()

        // Circular background
        draw.setColor(backgroundColor)
        draw.fillOval(-currentRadius, -currentRadius, currentRadius * 2, currentRadius * 2)

        draw.setColor(150, 150, 150)

        draw.setStroke(10f)
        draw.oval(-currentRadius, -currentRadius, currentRadius * 2, currentRadius * 2)

        draw.setStroke(3f)
        draw.oval(-currentRadius / 3, -currentRadius / 3, currentRadius * 2 / 3, currentRadius * 2 / 3)

        draw.setColor(200, 200, 200)

        draw.setStroke(5f)
        draw.oval(-currentRadius, -currentRadius, currentRadius * 2, currentRadius * 2)

        draw.setStroke(1f)
        draw.oval(-selectionRadius / 3,  -selectionRadius / 3, selectionRadius * 2 / 3, selectionRadius * 2 / 3)

        drawDragLine(draw)

        val itemRadius = currentRadius / (3 * (steps / 8.toFloat()))
        var marked = false

        selected = -1

        for (i in 0 until steps) {
            var active = false
            val x = -itemRadius + Math.sin((i / steps.toFloat()).toDouble() * Math.PI * 2.0).toFloat() * currentRadius
            val y = -itemRadius + Math.cos((i / steps.toFloat()).toDouble() * Math.PI * 2.0).toFloat() * currentRadius

            if (!marked && (Math.abs(pointerX) > selectionRadius / 3 || Math.abs(pointerY) > selectionRadius / 3)) {
                val pointer = ((Math.atan2((-pointerX).toDouble(), (-pointerY).toDouble()) / Math.PI + 1) / 2 * steps + 0.5f) % steps
                if (pointer.toInt() == i) {
                    marked = true
                    active = true
                }
            }

            val menuIndex = Math.floorMod(-i + count / 2, steps) // Makes items begin at 12 o'clock
            if (menuIndex < items.size) {
                val cmi = items[menuIndex]
                cmi.active = active
                cmi.radius = itemRadius
                cmi.translation.x = x
                cmi.translation.y = y

                if (active)
                    selected = menuIndex
            }
        }
    }

    private fun drawDragLine(draw: Draw) {
        draw.setColor(1.0f, 1.0f, 0f)
        draw.setStroke(5f)
        draw.line(0f, 0f, pointerX, pointerY)
        draw.fillOval(pointerX - 10f, pointerY - 10f, 20f, 20f)
    }

    fun setPointer(x: Float, y: Float) {
        pointerX = x
        pointerY = y
    }

    fun addMenuItem(item: ContextMenuItem) {
        items.add(item)
        add(item)
    }

    fun getSelected(): ContextMenuItem? {
        return if (selected > -1 && selected < items.size) items[selected] else null
    }

    fun animate() {
        currentRadius = 10f
    }
}
